drivers/ethernet/eth_smsc911x.c

```diff
@@ -0,0 +1,718 @@
+		return -1;
+	}
+
+	switch (((id >> 16) & 0xFFFF)) {
+	case 0x9220: /* SMSC9220 on MPS2 */
+	case 0x0118: /* SMS9118 as emulated by QEMU */
+		break;
+
+	default:
+		return -1;
+	}
+
+	return 0;
+}
+
+static int smsc_soft_reset(void)
+{
+	unsigned int time_out = RESET_TIMEOUT;
+
+	SMSC9220->HW_CFG |= HW_CFG_SRST;
+
+	do {
+		k_sleep(1);
+		time_out--;
+	} while (time_out != 0 && (SMSC9220->HW_CFG & HW_CFG_SRST));
+
+	if (time_out == 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+void smsc_set_txfifo(unsigned int val)
+{
+	/* 2kb minimum, 14kb maximum */
+	if (val >= 2 && val <= 14) {
+		SMSC9220->HW_CFG = val << 16;
+	}
+}
+
+void smsc_init_irqs(void)
+{
+	SMSC9220->INT_EN = 0;
+	/* Clear all interrupts */
+	SMSC9220->INT_STS = 0xFFFFFFFF;
+	/* Polarity config which works with QEMU */
+	/* IRQ deassertion at 220 usecs and master IRQ enable */
+	SMSC9220->IRQ_CFG = 0x22000111;
+}
+
+static int smsc_check_phy(void)
+{
+	u32_t phyid1, phyid2;
+
+	if (smsc_phy_regread(SMSC9220_PHY_ID1, &phyid1)) {
+		return -1;
+	}
+	if (smsc_phy_regread(SMSC9220_PHY_ID2, &phyid2)) {
+		return -1;
+	}
+	return ((phyid1 == 0xFFFF && phyid2 == 0xFFFF) ||
+	    (phyid1 == 0x0 && phyid2 == 0x0));
+}
+
+int smsc_reset_phy(void)
+{
+	u32_t val;
+
+	if (smsc_phy_regread(SMSC9220_PHY_BCONTROL, &val)) {
+		return -1;
+	}
+
+	val |= 1 << 15;
+
+	if (smsc_phy_regwrite(SMSC9220_PHY_BCONTROL, val)) {
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * Advertise all speeds and pause capabilities
+ */
+void smsc_advertise_caps(void)
+{
+	u32_t aneg_adv = 0;
+
+	smsc_phy_regread(SMSC9220_PHY_ANEG_ADV, &aneg_adv);
+	aneg_adv |= 0xDE0;
+
+	smsc_phy_regwrite(SMSC9220_PHY_ANEG_ADV, aneg_adv);
+	smsc_phy_regread(SMSC9220_PHY_ANEG_ADV, &aneg_adv);
+}
+
+void smsc_establish_link(void)
+{
+	u32_t bcr = 0;
+	u32_t hw_cfg = 0;
+
+	smsc_phy_regread(SMSC9220_PHY_BCONTROL, &bcr);
+	bcr |= (1 << 12) | (1 << 9);
+	smsc_phy_regwrite(SMSC9220_PHY_BCONTROL, bcr);
+	smsc_phy_regread(SMSC9220_PHY_BCONTROL, &bcr);
+
+	hw_cfg = SMSC9220->HW_CFG;
+	hw_cfg &= 0xF0000;
+	hw_cfg |= (1 << 20);
+	SMSC9220->HW_CFG = hw_cfg;
+}
+
+inline void smsc_enable_xmit(void)
+{
+	SMSC9220->TX_CFG = 0x2 /*TX_CFG_TX_ON*/;
+}
+
+void smsc_enable_mac_xmit(void)
+{
+	u32_t mac_cr = 0;
+
+	smsc_mac_regread(SMSC9220_MAC_CR, &mac_cr);
+
+	mac_cr |= (1 << 3);     /* xmit enable */
+	mac_cr |= (1 << 28);    /* Heartbeat disable */
+
+	smsc_mac_regwrite(SMSC9220_MAC_CR, mac_cr);
+}
+
+void smsc_enable_mac_recv(void)
+{
+	u32_t mac_cr = 0;
+
+	smsc_mac_regread(SMSC9220_MAC_CR, &mac_cr);
+	mac_cr |= (1 << 2);     /* Recv enable */
+	smsc_mac_regwrite(SMSC9220_MAC_CR, mac_cr);
+}
+
+int smsc_init(void)
+{
+	unsigned int phyreset = 0;
+
+	if (smsc_check_id() < 0) {
+		return -1;
+	}
+
+	if (smsc_soft_reset() < 0) {
+		return -1;
+	}
+
+	smsc_set_txfifo(5);
+
+	/* Sets automatic flow control thresholds, and backpressure */
+	/* threshold to defaults specified. */
+	SMSC9220->AFC_CFG = 0x006E3740;
+
+#if 0
```

Comment:
Get rid of this code if it's never meant to be compiled.

Reply:
> Get rid of this code if it's never meant to be compiled.

So, this driver is written for hardware which actually exists, even though it's probably will be mostly used for QEMU emulated (== simplified) hardware, and that's how I tested it (I don't have access to MPS2 board). So, I wouldn't like to cut (or complicate) way to make it run well on real hw too. The original mbedOS driver had this code, I didn't convert it as QEMU doesn't emulate EEPROM, but whenever possible, I'd prefer to leave it in the driver even if commented out.  I added comment why it is so.

Reply:
> So, I wouldn't like to cut (or complicate) way to make it run well on real hw too

Please do it then like this:
```
	/* Not needed on QEMU, but may be needed on real hardware. */
	if (!IS_ENABLED(CONFIG_QEMU_TARGET) && smsc_wait_eeprom()) {
		return -1;
	}
```

Reply:
> Please do it then like this:
>	if (!IS_ENABLED(CONFIG_QEMU_TARGET) && smsc_wait_eeprom()) 

That's confusing and misleading: not reading EEPROM doesn't depend on being QEMU_TARGET;  smsc_wait_eeprom() doesn't exit.

It's just a TODO item/note to someone else who may want to get that driver running on the real HW, it should be good as it is.

Reply:
> That's confusing and misleading: not reading EEPROM doesn't depend on being QEMU_TARGET; smsc_wait_eeprom() doesn't exit.

Your comment is then misleading as it says that calling smsc_wait_eeprom() is not needed on QEMU. Anyway, lets remove this piece of code that is never going to be called, it is just confusing everybody. If we run this with real hardware, then suitable spells will be added here anyway.

Reply:
> Anyway, lets remove this piece of code

Ok, done.